include/libint2/dfbs_generator.h

```diff
@@ -0,0 +1,210 @@
+
+        /// @brief computes the Coulomb matrix (\mu|rij^{-1}|\nu)  for a set of shells
+        /// @param shells set of shells
+        /// @return Coulomb matrix
+        Matrix compute_coulomb_matrix(const Shellvec &shells) {
+            const auto n = nbf(shells);
+            Matrix result = Matrix::Zero(n, n);
+            using libint2::Engine;
+            Engine engine(libint2::Operator::coulomb, max_nprim(shells), max_l(shells), 0);
+            engine.set(BraKet::xs_xs);
+            engine.set(ScreeningMethod::Conservative);
+            auto shell2bf = map_shell_to_basis_function(shells);
+            const auto &buf = engine.results();
+            for (auto s1 = 0; s1 != shells.size(); ++s1) {
+                auto bf1 = shell2bf[s1];
+                auto n1 = shells[s1].size();
+                for (auto s2 = 0; s2 <= s1; ++s2) {
+                    auto bf2 = shell2bf[s2];
+                    auto n2 = shells[s2].size();
+                    engine.compute(shells[s1], shells[s2]);
+                    Eigen::Map<const Matrix> buf_mat(buf[0], n1, n2);
+                    result.block(bf1, bf2, n1, n2) = buf_mat;
+                    if (s1 != s2)
+                        result.block(bf2, bf1, n2, n1) = buf_mat.transpose();
+                }
+            }
+            return result;
+        }
+
+        /// @brief Sorts a vector of shells by angular momentum
+        std::vector<Shellvec> sort_by_L(const Shellvec& shells){
+            int lmax = max_l(shells);
+            std::vector<Shellvec> sorted_shells;
+            sorted_shells.resize(lmax+1);
+            for(auto shell:shells){
+                auto l = shell.contr[0].l;
+                sorted_shells[l].push_back(shell);
+            }
+            return sorted_shells;
+        }
+
+    }// namespace detail
+
+    class DFBS_generator {
+    public:
+        /// @brief constructor for DFBS generator class, generates density fitting basis set from products of AO basis functions
+        /// @param obs_name name of basis set for AO functions
+        /// @param atoms vector of atoms
+        DFBS_generator(std::string obs_name,
+                       const std::vector<Atom> &atoms) : obs_name_(std::move(obs_name)), atoms_(std::move(atoms)) {
+            std::vector<Shellvec> obs_shell_vec;
+            std::vector<Shellvec> primitive_cluster;
+            // get AO basis shells for each atom
+            for (auto atom: atoms) {
+                auto atom_bs = BasisSet(obs_name_, {atom});
+                obs_shell_vec.emplace_back(atom_bs.shells());
+            }
+            // get primitive shells from AO functions
+            for (auto obs_shells: obs_shell_vec) {
+                primitive_cluster.emplace_back(uncontract(obs_shells));
+            }
+
+            //compute candidate shells
+            candidate_shells_ = datail::candidate_functions(primitive_cluster);
+
+            // initialize libint2
+            libint2::initialize();
```

Comment:
assert that libint is initialized already on entry, but no need to initialize here ... initialization is user's job, can't do it safely from within libint